src/hardware/siglent-sds/api.c

```diff
@@ -429,6 +437,36 @@ static int config_get(uint32_t key, GVariant **data,
 			*data = g_variant_new_string("History");
 		break;
 	case SR_CONF_SAMPLERATE:
+		/* SDS2000X+: The channels need to be enabled/disabled prior to
+		 * obtaining the sample rate when the data source is set to display.
+		 * If a channel is turned on/off on the oscilloscope then use Sigrok's settings instead. */
+		if (!strcmp(devc->model->series->name, "SDS2000X+") && devc->data_source == DATA_SOURCE_SCREEN) {
+			GSList *l;
+			for (l = sdi->channels; l; l = l->next) {
+				ch = l->data;
+				if (ch->type == SR_CHANNEL_ANALOG) {
+					char *channel_state;
+					char *cmd = g_strdup_printf("C%d:TRA?", ch->index + 1);
+					sr_scpi_get_string(sdi->conn, cmd, &channel_state);
+					if (!strcmp(channel_state, "ON") && !ch->enabled)
+						if (siglent_sds_config_set(sdi, "C%d:TRA %s", ch->index + 1,
+							ch->enabled ? "ON" : "OFF") != SR_OK)
+							return SR_ERR;
+					if (!strcmp(channel_state, "OFF") && ch->enabled)
+						if (siglent_sds_config_set(sdi, "C%d:TRA %s", ch->index + 1,
+							ch->enabled ? "ON" : "OFF") != SR_OK)
+							return SR_ERR;
```

Comment:
this is getting into some pretty deep indent levels, maybe time to split some functionality into a separate func ?